src/parser/jobs/blm/modules/RotationWatchdog.tsx

```diff
@@ -0,0 +1,563 @@
+	[ERROR_CODES.MANAFONT_BEFORE_DESPAIR]: <Trans id="blm.rotation-watchdog.error-messages.manafont-before-despair"><ActionLink {...ACTIONS.MANAFONT}/> used before <ActionLink {...ACTIONS.DESPAIR}/></Trans>,
+	[ERROR_CODES.MISSING_DESPAIRS]: <Trans id="blm.rotation-watchdog.error-messages.missing-despair">Rotation didn't include <ActionLink {...ACTIONS.DESPAIR}/></Trans>,
+	[ERROR_CODES.NO_FIRE_SPELLS]: <Trans id="blm.rotation-watchdog.error-messages.no-fire-spells">Rotation included no Fire spells</Trans>,
+	[ERROR_CODES.EXTRA_F1]: <Trans id="blm.rotation-watchdog.error-messages.extra-f1">Extra <ActionLink {...ACTIONS.FIRE_I}/></Trans>,
+}
+
+class Cycle {
+	casts: any[] = []
+	startTime: number
+	endTime?: number
+
+	hasManafont: boolean = false
+
+	inFirePhase: boolean = false
+	atypicalAFStart: boolean = false
+	firePhaseStartMP: number = 0
+
+	firstCycle: boolean = false
+	finalOrDowntime: boolean = false
+
+	gaugeStateBeforeFire: GaugeState = new GaugeState()
+
+	_errorCode: number = 0
+	public set errorCode(code: number) {
+		this._errorCode = Math.max(this._errorCode, code)
+	}
+	public get errorCode() {
+		return this._errorCode
+	}
+
+	/**
+	 * Greatly simplified from the pre-Shadowbringers version of this function. Because Umbral Soul gives us
+	 * a proper downtime action to build UH/UI, it should be possible to enter every fire phase with a normal
+	 * gauge state, so we're no longer going to reduce the expected count based on the actual gauge state on
+	 * entering fire phase. We're still tracking the data necessary to do this again in the future if there is
+	 * value in doing so.
+	 */
+	public get expectedFire4s(): number | undefined {
+		if (this.finalOrDowntime) {
+			return
+		}
+		// Account for the no-UH opener when determining the expected count of Fire 4s
+		return (this.firstCycle && this.gaugeStateBeforeFire.umbralHearts === 0 ?
+			NO_UH_OPENER_FIRE4 : EXPECTED_FIRE4) + (this.hasManafont ? FIRE4_FROM_MANAFONT : 0)
+	}
+	public get actualFire4s(): number {
+		return this.casts.filter(cast => cast.ability.guid === ACTIONS.FIRE_IV.id).length
+	}
+	public get missingFire4s(): number | undefined {
+		if (!this.expectedFire4s) { return }
+		return this.expectedFire4s - this.actualFire4s
+	}
+
+	public get expectedDespairs(): number {
+		return this.hasManafont ? 2 : 1
+	}
+	public get actualDespairs(): number {
+		return this.casts.filter(cast => cast.ability.guid === ACTIONS.DESPAIR.id).length
+	}
+	public get missingDespairs(): number {
+		return this.expectedDespairs - this.actualDespairs
+	}
+
+	constructor(start: number, gaugeState: GaugeState, isFirst: boolean = false) {
+		this.startTime = start,
+		// Object.assign because this needs to be a by-value assignment, not by-reference
+		this.gaugeStateBeforeFire = Object.assign(this.gaugeStateBeforeFire, gaugeState)
+		this.firstCycle = isFirst
+	}
+}
+
+// TS typedef for BLM Gauge events so it doesn't choke
+interface BLMGaugeEvent extends Event {
```

Comment:
might be worth tossing a comment mentioning it should be moved if/when it gets translated